wavelength-plugins/lottery.js

```diff
@@ -0,0 +1,120 @@
+ * Lottery Plug-in for Pokémon Showdown
+ *            Created by:
+ *         HoeenHero and Insist
+ ****************************************/
+
+"use strict";
+
+class Lottery {
+	constructor(room, user) {
+		this.players = [];
+		this.room = room;
+		room.lottoNumber = room.lottoNumber ? room.lottoNumber++ : 1;
+		this.lottoNumber = room.lottoNumber;
+		this.costToJoin = 3;
+		this.room.add(`|uhtml|lottery-${this.lottoNumber}|<div class="broadcast-blue"><p style="font-size: 14pt; text-align: center">A new <strong>Lottery drawing</strong> is starting!</p><p style="font-size: 9pt; text-align: center"><button name="send" value="/lotto join">Join</button><br /><strong>DISCLAIMER: Joining costs ${this.costToJoin} ${currencyPlural}!!!!</strong></p></div>`, true);
+		this.timer = setTimeout(() => {
+			if (this.players.length < 1) {
+				this.room.add('|uhtmlchange|lottery-' + this.lottoNumber + '|<div class="broadcast-red"><p style="text-align: center; font-size: 14pt>This Lottery drawing has ended due to lack of users.</p></div>');
+				return this.end();
+			}
+			this.drawWinner();
+		}, 1000 * 60 * 60 * 24);
+	}
+
+	onConnect(user, connection, room) {
+		user.sendTo(this.room, '|uhtml|lottery-' + this.lottoNumber + '|<div class="broadcast-blue"><p style="text-align: center; font-size: 14pt>A Lottery Drawing has started looking for players!<hr /><br />For the price of 3 ' + currencyPlural + ', you can earn 5 ' + currencyPlural + ' plus one ' + currencyName + ' per user who joins.</p><br /><button name="send" value="/lottery join">Click here to join the Lottery</button></div>');
+	}
+
+	drawWinner() {
+		let winner = this.players[Math.floor(Math.random() * this.players.length)];
+		let basePrize = 5;
+		let lottoPrize = basePrize + this.players.length + this.costToJoin;
+		this.room.add(`|html|<div class="infobox"><center><strong>Congratulations</strong> ${WL.nameColor(Users(winner), true)}!!! You have won the reward of ${lottoPrize} ${currencyPlural}</center></div>`);
+		Economy.writeMoney(winner, lottoPrize);
+		Economy.logTransaction(`${winner} has won the Lottery prize of ${lottoPrize} ${currencyPlural}`);
+		this.end();
+	}
+
+	joinLottery(user) {
+		Economy.readMoney(user.userid, money => {
+			if (this.players.includes(user)) return user.sendTo(this.room, "You have already joined this Lottery drawing.");
+			if (money < this.costToJoin) {
+				user.sendTo(this.room, 'You do not have enough ' + currencyPlural + ' to join.');
+				return;
+			}
+			Economy.writeMoney(user.userid, -this.costToJoin, () => {
+				Economy.readMoney(user.userid, money => {
+					Economy.logTransaction(user.name + " entered a Lottery drawing for " + this.costToJoin + " " + currencyPlural + ".");
+				});
+			});
+			this.players.push(user);
+		});
+	}
+
+	leaveLottery(user) {
+		Economy.writeMoney(user.userid, this.costToJoin, () => {
+			if (!this.players.includes(user)) return user.sendTo(this.room, `You are not currently in the Lottery drawing in this room..`);
```

Comment:
This is a huge exploit, check to make sure they arent in the lottery before refunding.

Reply:
Ok

Reply:
Still not patched

Reply:
Like they can't gain money with that being how it is o3o

Reply:
`Economy.writeMoney(user.userid, this.costToJoin, () => {`
user was refuned
`if (!this.players.includes(user)) return user.sendTo(this.room, `You are not currently in the Lottery drawing in this room..`);`
User was told they haven't joined and cant be refunded.

See the issue? Check if they joined, and if not error out BEFORE calling `Economy.writeMoney`

Reply:
I thought I errored it out beforehand with the new update?

Reply:
Fine ill do it myself